src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+            }
+
+            _consumed++;
+            _lineBytePosition = 0;
+            _lineNumber++;
+        Done:
+            _consumed += 2 + idx;
+            return true;
+        }
+
+        private bool SkipMultiLineComment(ReadOnlySpan<byte> localCopy, out int idx)
+        {
+            idx = 0;
+            while (true)
+            {
+                int foundIdx = localCopy.Slice(idx).IndexOf(JsonConstants.Solidus);
+                if (foundIdx == -1)
+                {
+                    if (IsLastSpan)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.EndOfCommentNotFound);
+                    else
+                        return false;
+                }
+                if (foundIdx != 0 && localCopy[foundIdx + idx - 1] == '*')
+                {
+                    idx += foundIdx;
+                    break;
+                }
+                idx += foundIdx + 1;
+            }
+
+            Debug.Assert(idx >= 1);
+            _consumed += 3 + idx;
+
+            (int newLines, int newLineIndex) = JsonReaderHelper.CountNewLines(localCopy.Slice(0, idx - 1));
+            _lineNumber += newLines;
+            if (newLineIndex != -1)
+                _lineBytePosition = idx - newLineIndex;
+            else
+                _lineBytePosition += 4 + idx - 1;
+            return true;
+        }
+
+        private bool ConsumeComment()
+        {
+            //Create local copy to avoid bounds checks.
+            ReadOnlySpan<byte> localCopy = _buffer.Slice(_consumed + 1);
+
+            if (localCopy.Length > 0)
+            {
+                byte marker = localCopy[0];
+                if (marker == JsonConstants.Solidus)
+                    return ConsumeSingleLineComment(localCopy.Slice(1));
+                else if (marker == '*')
+                    return ConsumeMultiLineComment(localCopy.Slice(1));
+                else
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedStartOfValueNotFound, marker);
+            }
+
+            if (IsLastSpan)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedStartOfValueNotFound, JsonConstants.Solidus);
+            else
+                return false;
+
+            return true;
+        }
+
+        private bool ConsumeSingleLineComment(ReadOnlySpan<byte> localCopy)
+        {
+            if (!SkipSingleLineComment(localCopy, out int idx))
+                return false;
+
+            ValueSpan = localCopy.Slice(0, idx);
+            _stack.Push(_tokenType);
+            _tokenType = JsonTokenType.Comment;
+            return true;
+        }
+
+        private bool ConsumeMultiLineComment(ReadOnlySpan<byte> localCopy)
+        {
+            if (!SkipMultiLineComment(localCopy, out int idx))
+                return false;
+
+            ValueSpan = localCopy.Slice(0, idx - 1);
+            _stack.Push(_tokenType);
```

Comment:
Doesn't the stack mainly track _isObject vs _isArray?  

```
{
    "value": 11,
    /* yes, it's mis-spelled */
    "deelay": 3
}
```

or

```
[
    12,
    87,
    /* Isn't it "nice" that JSON provides no limits on the length of numbers? */
    123456789012345678901234567890123456789.01234567890123456789e+9876543218976543219876543210
]
```

Feels like it would push "Number", if I followed the state diagram correctly; but the calls to Pop seem to just be from EndArray/EndObject.

I don't see a reason why the stack and comments interact, since (if I understood correctly) the stack and numbers don't.

Reply:
Seems like it's not really a stack, just a second JsonTokenType field in your state struct.  (Where one of them (which gets set to Comment here) is "what does TokenType say", only read by the TokenType property; the other is "what state am I in", which gets set everywhere except here.)